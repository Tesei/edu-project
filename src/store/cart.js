import { defineStore } from 'pinia'
import { reactive, computed, ref } from 'vue'
import { useGoodsStore } from '@/store/goods'
const goodsStore = useGoodsStore()
import axios from 'axios'

export const useCartStore = defineStore(
    'cart',
    () => {
        const cartList = ref({})
        const getCartList = computed(() => cartList.value)

        function addToCartList(item) {
            if (cartList.value[item.title]) cartList.value[item.title]++
            else cartList.value[item.title] = 1
        }
        function removeFromCartList(item) {
            if (cartList.value[item.title] && cartList.value[item.title] > 1) cartList.value[item.title]--
            else if (cartList.value[item.title] && cartList.value[item.title] <= 1) delete cartList.value[item.title]
            else return
        }

        function cleanItemFromBucket(item) {
            delete cartList.value[item.title]
        }
        function cleanBucket() {
            Object.keys(cartList.value).forEach((key) => delete cartList.value[key])
            hasDelivery.value = false
        }

        const getCountOfGoodsNames = computed(() => {
            return Object.keys(cartList.value).length
        })

        const getCountOfGoodsItems = computed(() => {
            return Object.entries(cartList.value).reduce((acc, value) => acc + value[1], 0)
        })

        const getTotalSumOfGoods = computed(() => {
            return Object.entries(cartList.value).reduce((acc, value) => {
                let priceItem = goodsStore.goods.find((item) => item.title === value[0])?.price
                if (typeof priceItem === 'string') priceItem = priceItem.replace(/[^0-9.]/gim, '')
                return +priceItem * value[1] + acc
            }, 0)
        })
        const getTotalSumOfGoodsAndService = computed(() => {
            return hasDelivery.value ? getTotalSumOfGoods.value + 1000 : getTotalSumOfGoods.value
        })

        const hasDelivery = ref(false)
        function changeDelivery() {
            hasDelivery.value = !hasDelivery.value
        }

        const showClientForm = ref(false)
        function changeVisibleClientForm() {
            showClientForm.value = !showClientForm.value
        }

        async function sendBuyForm(form) {
            try {
                let res = await axios.get('https://httpbin.org/', {
                    timeout: 4000,
                    data: form,
                })
                if (res.status >= 200 && res.status < 300) return 'success'
                else throw new Error('bad status')
            } catch {
                console.error('Ошибка в методе sendBuyForm')
            }
        }

        const savedCartLists = reactive([])
        function saveUserData(userId) {
            const data = {
                id: userId,
                cartList: { ...cartList.value },
            }
            savedCartLists.push(data)
            cleanBucket()
        }
        function loadSavedUserData(userId) {
            let newData = savedCartLists.find((user) => user.id === userId)
            if (newData) cartList.value = { ...newData.cartList }
        }

        return {
            addToCartList,
            getCartList,
            removeFromCartList,
            getCountOfGoodsNames,
            getCountOfGoodsItems,
            getTotalSumOfGoods,
            cleanBucket,
            cleanItemFromBucket,
            changeDelivery,
            hasDelivery,
            getTotalSumOfGoodsAndService,
            showClientForm,
            changeVisibleClientForm,
            sendBuyForm,
            saveUserData,
            loadSavedUserData,
            savedCartLists,
            cartList,
        }
    },
    { persist: true },
)
