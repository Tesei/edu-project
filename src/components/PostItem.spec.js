import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import component from './PostItem.vue'
import { createTestingPinia } from '@pinia/testing'
import { useCartStore } from '@/store/cart.js'

import flushPromises from 'flush-promises'
import waitForExpect from 'wait-for-expect'

describe('PostItem component', () => {
    
    const buildWrapper = (stubActions = true) => {
        return mount(component, {
            props: {
                post: {
                    'id': 1,
                    'title': 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
                    'price': 1009.95,
                    'description': 'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
                    'category': 'men\'s clothing',
                    'image': 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
                    'rating': {
                        'rate': 3.9,
                        'count': 120
                    }
                },
            },
            global: {
                plugins: [createTestingPinia({ stubActions })],
            },
        })
    }

    it('mounts without error', () => {
        const wrapper = buildWrapper()
        expect(wrapper.exists).toBeTruthy()
        expect(wrapper.html()).toMatchSnapshot()
    })
    it('start btn text', () => {
        const wrapper = buildWrapper()
        expect(wrapper.find('.card__btn').text()).toEqual('В корзину')
    })
    it('price text has a space', () => {
        const wrapper = buildWrapper()
        const price = 1009.95.toLocaleString('ru-RU') + ' руб.'
        expect(wrapper.find('.card__price').text()) === price
    })
    it('emit called', async () => {
        const wrapper = buildWrapper()
        const btn = wrapper.find('.card__btn')
        await btn.trigger('click')
        await btn.trigger('click')
        await btn.trigger('click')
        expect(wrapper.emitted()['click-bye']).toHaveLength(3)
    })
    it('function addToCartList been called', async () => {
        const wrapper = buildWrapper(false)
        const store = useCartStore()     
        await store.addToCartList(wrapper.props().post)
        await flushPromises() 
        await waitForExpect(() => {
            expect(store.addToCartList).toHaveBeenCalledTimes(1)
        })
    })
})
