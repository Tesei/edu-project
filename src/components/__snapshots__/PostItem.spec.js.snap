// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostItem component > mounts without error 1`] = `
"<article data-v-cf937a3e="" class="goods__item card">
  <!-- <div
			class="card__bucket"
			@click="$emit('remove', post)"
		>
			<div class="card__bucket-picture">
				<img
					src="@/assets/images/icons/bucket-trash.svg"
					alt="Удалить товар"
					class="card__bucketImage"
				>
			</div>
		</div> -->
  <div data-v-cf937a3e="" class="card__content">
    <div data-v-cf937a3e="" class="card__picture _ibg"><img data-v-cf937a3e="" src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg" alt="Изображние товара" class="card__img"></div>
    <div data-v-cf937a3e="" class="card__about"><span data-v-cf937a3e="" class="card__title">Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops</span><span data-v-cf937a3e="" class="card__text">Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday</span><span data-v-cf937a3e="" class="card__price">1&nbsp;009,95<span data-v-cf937a3e=""> руб.</span></span>
      <div data-v-43357113="" data-v-cf937a3e="" class="button card__btn"><button data-v-43357113="" type="submit" class="btn _active">В корзину</button></div>
    </div>
  </div>
</article>"
`;
